controlplane/kubeadm/controllers/remediation.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controlplane/kubeadm/internal"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// reconcileUnhealthyMachines handles KubeadmControlPlane unhealthyMachine reconciliation.
+func (r *KubeadmControlPlaneReconciler) reconcileUnhealthyMachines(ctx context.Context, c *internal.ControlPlane) (ret ctrl.Result, retErr error) {
+	logger := r.Log.WithValues("namespace", c.KCP.Namespace, "kubeadmControlPlane", c.KCP.Name, "cluster", c.Cluster.Name)
+
+	// Gets all machines that have a MachineHealthCheckSucceeded condition set to False,
+	// indicating a problem was detected on the machine, and the MachineOwnerRemediated condition set, indicating that KCP is
+	// responsible of performing remediation as owner of the machine.
+	unhealthyMachines := c.UnhealthyMachines()
+
+	// If there are not unhealthy machines, return so KCP can proceed with other operations (ctrl.Result nil).
+	if len(unhealthyMachines) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// Select the machine to be remediated, which is the oldest machine marked as unhealthy.
+	// NOTE: The current solution is considered acceptable for the most frequent use case (only one unhealthy machine),
+	// however, in the future this could potentially be improved for the scenario where more than one unhealthy machine exists
+	// by considering which machine has lower impact on etcd quorum.
```

Comment:
Let's track this in an issue, and schedule the work for v1alpha4?

Reply:
kk, I will open the issue. Not sure about the priority, given that this effort provides benefits only when
- two CP nodes are marked unhealthy 
- another machine has a failing etcd member

Reply:
see https://github.com/kubernetes-sigs/cluster-api/issues/3845